### PR TITLE
Use `llvm-readelf` to detect and bundle needed dynamic libs for Android

### LINKED ISSUE #47: use `llvm-readelf` to figure out dynamic libraries to include

### PR DESCRIPTION
**Pull request by MarijnS95**
Fixes #47

Dynamic libraries that a Rust library links against while not being present on the target platform (`libc++_shared.so` and other custom libraries) need to be bundled in the APK specifically.  This approach is adopted frm `android-ndk-rs` while switching from `readelf` to `llvm-readelf`.

### CODE REVIEW

File: xbuild/src/command/build.rs
Comment (MarijnS95) [inline xbuild/src/command/build.rs:274]: @dvc94ch Do you remember what was up with this? We can't call `foo.join(..).as_path()` since the temporary owned `PathBuf` created by `fn join()` would be dropped here, but somehow that isn't the case with `&*`??

File: xbuild/src/lib.rs
Comment (MarijnS95) [inline xbuild/src/lib.rs:299]: `x86_64` was missing here, so I figured we could just reuse `ndk_triple` since they're the same?

File: xcommon/src/llvm.rs
Comment (MarijnS95) [inline xcommon/src/llvm.rs:71]: Still need to test this...

File: xbuild/src/lib.rs
Comment (dvc94ch) [reply xbuild/src/lib.rs:299]: well, if we add more targets it might not be. I'd prefer to not reuse `ndk_triple` here

File: xbuild/src/command/build.rs
Comment (dvc94ch) [reply xbuild/src/command/build.rs:270]: I guess it's probably easier to review the entire "dynamic library" handling at once? Also need some kind of test/example so we can check that it works for all platforms in ci.

File: xbuild/src/command/build.rs
Comment (dvc94ch) [reply xbuild/src/command/build.rs:274]: apparently

File: xcommon/src/llvm.rs
Comment (dvc94ch) [reply xcommon/src/llvm.rs:71]: trimming it first can't hurt

**Comment by dvc94ch** [comment]
not sure, will have to try. at least for ios we don't need to bother, since rust doesn't support building dynamic libraries for ios. and I really don't have the time to mess about with major change proposals.

https://github.com/rust-lang/rust/pull/95847

File: xbuild/src/lib.rs
Comment (MarijnS95) [reply xbuild/src/lib.rs:299]: I'll add the missing `x86_64-linux-android` arch separately then.

File: xbuild/src/command/build.rs
Comment (MarijnS95) [reply xbuild/src/command/build.rs:274]: I trust it's safe since the borrow checker isn't complaining, but it still has me bamboozled.

Comment (MarijnS95) [reply xbuild/src/command/build.rs:274]: https://doc.rust-lang.org/stable/reference/destructors.html#temporary-lifetime-extension Found it, it's relying on scope extension for temporaries. The scope for the results of `join()` are extended to the surrounding scope of the `let` binding, instead of remaining within the expression (`[...]` array constructor) of said `let` binding.

Note that it also/only works with just a `&` on the `PathBuf`, `Deref`ing into `Path`. It doesn't work through `as_path()` since I guess Rust "can't easily follow" what's happening inside that function and as such know that it would have to prolong the lifetime of `self`.

The `*` is just added to force a deref into `Path` first, before turning it into `&Path` again. This can be omitted if the first item is already an `&Path` or the `let` binding would be explicitly annotated to be `&Path` _and not `&PathBuf`.

The more you know :grin:

File: xcommon/src/llvm.rs
Comment (MarijnS95) [reply xcommon/src/llvm.rs:71]: It's gonna be a bunch of `.trim()` calls, but sure.

Comment (MarijnS95) [reply xcommon/src/llvm.rs:71]: Fwiw `fn lines()` seems to be implemented as `split_terminator('\n')` which doesn't take `\r\n` into account?

File: xbuild/src/command/build.rs
Comment (MarijnS95) [reply xbuild/src/command/build.rs]: I implemented the missing bits, will see about some CI action to test this functionality next.

**Comment by MarijnS95** [comment]
> not sure, will have to try. at least for ios we don't need to bother, since rust doesn't support building dynamic libraries for ios. and I really don't have the time to mess about with major change proposals.

Thanks, as I'm not sure what you can do in terms of the other packaging targets x supports. Will probably need some additional refactoring though :/

File: xbuild/src/cargo/mod.rs
Comment (dvc94ch) [reply xbuild/src/cargo/mod.rs:107]: the reason for that is we set env variables for different targets, causing cargo to think the build is out of date when compiling for a different target. this way cargo caches the builds properly

Comment (MarijnS95) [reply xbuild/src/cargo/mod.rs:107]: We've also had slowdowns with this with `cargo-apk` when building Android and host-native binaries in the same `target/` even though everything non-host ends up in `target/<triple>` and shouldn't touch `target/`, you're saying that the env vars are the reason for these seemingly unnecessary rebuilds?

Comment (dvc94ch) [reply xbuild/src/cargo/mod.rs:107]: yes, cargo stores the env vars as inputs to the build globally, not per target. if it were stored per target as long as you always use `x`, we wouldn't have to override the `target-dir`. but I'd still put it in `target/x/target`, to avoid accidentally running `cargo build` and then having to rebuild from scratch wenn running `x build`.

**Comment by MarijnS95** [comment]
```
--------------------clang/llvm toolchain--------------------
clang                version             /usr/bin/clang
clang++              version             /usr/bin/clang++
llvm-ar              not found
llvm-lib             not found
llvm-readelf         not found
lld                  not found
lld-link             not found
lldb                 not found
lldb-server          not found
```

@dvc94ch Turns out GH actions CI doesn't ship with `llvm-readelf`, and I still don't like the ``Failed to run `the command` `` error with only a `file not found` message, as that really doesn't cover the load and may be interpreted as individual arguments being missing.

**Comment by dvc94ch** [comment]
should probably be llvm-readobj (llvm-readelf is just a symlink). you should setup ci like done by windows actions.

**Comment by dvc94ch** [comment]
Can't solve all problems at the same time. The clang/lldb/java/javac binaries are going to have to be preinstalled for the foreseeable future. It's possible we could try using rust-lld and llvm tools from rustup. Although it looks like it could disappear at any time since it's still a preview feature

**Comment by MarijnS95** [comment]
> should probably be llvm-readobj (llvm-readelf is just a symlink). you should setup ci like done by windows actions.

Oh okay, I thought it was already there. Will double-check the script and setup.

**Comment by MarijnS95** [comment]
So it seems I've been reading the CI results for at least https://github.com/cloudpeers/xbuild/runs/6261865757?check_suite_focus=true all wrong.

1. LLVM is installed on Windows already (https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md#language-and-runtime), and has the same files as the GH-actions installer, so that's not necessary (@dvc94ch how about the other platforms?);
2. The build succeeds there for Windows, with `llvm-readobj`. I've checked and the Windows installer indeed only provides `llvm-readobj`, not the `readelf` symlink (sensibly so, Windows doesn't use ELF);
3. The only build failing there is Ubuntu... Does it _not_ have `llvm-readobj` but only `llvm-readelf`??

**Comment by dvc94ch** [comment]
What's why are you symlinking llvm-readobj-12? Maybe you're trying to run a dangling symlink?

**Comment by MarijnS95** [comment]
> What's why are you symlinking llvm-readobj-12? Maybe you're trying to run a dangling symlink?

Shall I add a comment for these? Ubuntu only provides `llvm-read{elf,obj}-1{0,1,2}` so it won't be found when looking for `llvm-readobj`. Apparently `llvm-readelf` wasn't previously found on Ubuntu builds either.

**Comment by MarijnS95** [comment]
Looks like the CI succeeds now, thanks to the symlinks :partying_face: 

@dvc94ch I'll add a test for a small C++ file through the `cc` crate that requires `libc++_shared.so` on Android (as a third example?) then I think this is ready for a final round of review!

File: xcommon/src/llvm.rs
Comment (dvc94ch) [inline xcommon/src/llvm.rs:49]: yep, this needs to go. you can compute the blocklist for android by searching the provided_libs and filtering for `libc++_shared.so`.

File: xbuild/src/command/build.rs
Comment (dvc94ch) [inline xbuild/src/command/build.rs:324]: the build.rs file is getting quite large. maybe we can have a file for each platform handling the final packaging.

**Review by dvc94ch — COMMENTED** [review]
Looking good, thanks for working on this!

File: xcommon/src/llvm.rs
Comment (MarijnS95) [reply xcommon/src/llvm.rs:17]: That sounds like a lot of work to update and maintain?

Comment (MarijnS95) [reply xcommon/src/llvm.rs:17]: Also if we don't update this with new libraries, the logic will just copy the new libraries even when they are already available on the phone, making an APK unnecessarily bigger than it needs to be.

Comment (MarijnS95) [reply xcommon/src/llvm.rs:17]: Note that none of the items in `provided_libs_paths` is used to copy or blacklist anything at all. That only happens for `search_paths`. The goal of `provided_libs_paths` is to at least validate that a library is available (without copying it!) so that we can otherwise provide an error instead of having the app fail at runtime.

File: xbuild/src/command/build.rs
Comment (MarijnS95) [reply xbuild/src/command/build.rs:324]: Yes, that's what I was hinting at when mentioning it being hard to find the right pieces of code to edit.

I don't want to do that as part of this PR though to make it unnecessarily larger. Feel free to pick it up after merging, or before and I'll rebase this on top.

File: xcommon/src/llvm.rs
Comment (MarijnS95) [reply xcommon/src/llvm.rs:49]: And then I copypaste that to a list somewhere... And document the manual commands/steps someone needs to go through to regenerate that list? Please no :(

Comment (MarijnS95) [reply xcommon/src/llvm.rs:49]: Note also that your change implies that we'll add all the Android library paths from `provided_libs_paths` to `search_paths` which pretty much defeat the purpose since there's exactly one library from there that we need to copy, the rest should not be copied but only used to validate that all dynamic-link dependencies are met.

File: xcommon/src/llvm.rs
Comment (dvc94ch) [reply xcommon/src/llvm.rs:17]: for linux we already have an excludelist [0]. for android we can compute the exclude list by finding all libraries in provided_libs_path - libc++_shared.so. if it's in the exclude list we skip it otherwise we use search_paths to locate it and include it. for macos/windows we can compute the exclude list similarly.

- [0] https://github.com/AppImage/pkg2appimage/blob/master/excludelist

Comment (MarijnS95) [reply xcommon/src/llvm.rs:17]: In this sense `provided_libs_paths` already is a blocklist, we just need to remove `libc++_shared.so`. Are you asking for this list to be checked in to the repo somewhere, or simply to refactor this so that it's still computed at runtime but more clear in the code? I can do the latter :)

File: xcommon/src/llvm.rs
Comment (dvc94ch) [reply xcommon/src/llvm.rs:49]: provided_libs_paths must already be included in search_paths. the search paths are anything added via `-L`, otherwise there's a bug.

File: xcommon/src/llvm.rs
Comment (dvc94ch) [reply xcommon/src/llvm.rs:17]: well, it contains the library name not the path. the block list would be all file names with an so extension within provided paths. so something like `filename(find search_paths | grep *.so)`

Comment (dvc94ch) [reply xcommon/src/llvm.rs:17]: in a sense it's already an excludelist, was just  suggesting how to generalize it. it requries a bit more work but should be more generic.

**Review by dvc94ch — APPROVED** [review]

File: xcommon/src/llvm.rs
Comment (MarijnS95) [reply xcommon/src/llvm.rs:49]: They are not, and it should be by design?

File: xcommon/src/llvm.rs
Comment (MarijnS95) [reply xcommon/src/llvm.rs:17]: I think you have a fair point yes, but as pointed out below perhaps you weren't aware that `provided_libs_paths` is *not* included in `search_paths` currently.

Perhaps we can improve that as I quite like your suggestion. It is indeed already converted into an exclude list, with `libc++_shared.so` removed from it:

https://github.com/cloudpeers/xbuild/blob/30b0849d20b22825b233d19245a3f1c808e787dc/xcommon/src/llvm.rs#L22-L29

Comment (MarijnS95) [reply xcommon/src/llvm.rs:17]: > ah, this is good enough for now, let's just merge it.

Your call, we can also merge it now and improve in the future. I don't mind either way, I think we just misunderstood each others ideas in the beginning :)

File: xbuild/src/command/build.rs
Comment (MarijnS95) [reply xbuild/src/command/build.rs:268]: @dvc94ch I checked this and `has_lib` is false, that's why this bit of code never runs.

Comment (MarijnS95) [reply xbuild/src/command/build.rs:268]: There's a `libapp.so` but I don't see what `main.rs` compiled into...

File: xcommon/src/llvm.rs
Comment (dvc94ch) [reply xcommon/src/llvm.rs:49]: well, maybe it's just named confusingly. but if the search path is the same that is passed to the linker, then any library that was linked must be in the search path. So this means that `libc++_shared.so` would be in both the search path and the provided path. So the provided path seems like it should be a subset of search path.

Comment (dvc94ch) [reply xcommon/src/llvm.rs:49]: libc++ is a poor example, since we actually want to remove it from the provided path.

File: xbuild/src/command/build.rs
Comment (dvc94ch) [reply xbuild/src/command/build.rs:268]: fyi adding the template to ci https://github.com/cloudpeers/xbuild/pull/71

Comment (MarijnS95) [reply xbuild/src/command/build.rs:268]: Okay, built it and `has_lib` is true, I see `libflutter.so` being added but it's not in the required dylibs:

```console
$ llvm-readelf --needed-libs target/x/release/android/arm64/cargo/aarch64-linux-android/release/libfoo.so
NeededLibraries [
  libc.so
  libdl.so
]
```

Comment (dvc94ch) [reply xbuild/src/command/build.rs:268]: ah, makes sense. it's probably dlopened by the flutter java code. sorry for the false alarm. In that case this is resolved.

Comment (MarijnS95) [reply xbuild/src/command/build.rs:268]: > you should not have to install dart. it is already prepackaged and shipped. what os are you trying this on?

Arch linux. It did download things but I think `dart` wasn't made available to the path properly? Hopefully the CI will catch it and allow you to debug, as it wasn't an issue in `helloworld` with just `main.rs`.

> fyi adding the template to ci #71

Thanks, nice! I'll rebase on that when it's merged so that we can be doubly sure that this PR is not breaking Flutter.

> ah, makes sense. it's probably dlopened by the flutter java code. sorry for the false alarm. In that case this is resolved.

Yeah, I guess it must be dlopened then. No worries, better safe than sorry!

Comment (dvc94ch) [reply xbuild/src/command/build.rs:268]: can you let me know when that happened? was it during the build or when trying to run?

Comment (MarijnS95) [reply xbuild/src/command/build.rs:268]: Yup, it's `FfiGen` in `build.rs`:

<details>

```
   Compiling foo v0.1.0 (xbuild/examples/foo)
error: failed to run custom build command for `foo v0.1.0 (xbuild/examples/foo)`

Caused by:
  process didn't exit successfully: `xbuild/examples/foo/target/x/release/android/arm64/cargo/release/build/foo-09500578752ec7e1/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=xbuild/examples/foo/api.rsh

  --- stderr
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: dart not installed

  Caused by:
      No such file or directory (os error 2)

  Stack backtrace:
     0: <E as anyhow::context::ext::StdError>::ext_context
     1: anyhow::context::<impl anyhow::Context<T,E> for core::result::Result<T,E>>::context::{{closure}}
     2: core::result::Result<T,E>::map_err
     3: anyhow::context::<impl anyhow::Context<T,E> for core::result::Result<T,E>>::context
     4: ffi_gen::FfiGen::generate_dart
     5: build_script_build::main
     6: core::ops::function::FnOnce::call_once
     7: std::sys_common::backtrace::__rust_begin_short_backtrace
     8: std::rt::lang_start::{{closure}}
     9: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/ops/function.rs:280:13
    10: std::panicking::try::do_call
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:492:40
    11: std::panicking::try
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:456:19
    12: std::panic::catch_unwind
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panic.rs:137:14
    13: std::rt::lang_start_internal::{{closure}}
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:128:48
    14: std::panicking::try::do_call
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:492:40
    15: std::panicking::try
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:456:19
    16: std::panic::catch_unwind
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panic.rs:137:14
    17: std::rt::lang_start_internal
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:128:20
    18: std::rt::lang_start
    19: main
    20: __libc_start_call_main
    21: __libc_start_main@GLIBC_2.2.5
    22: _start', build.rs:12:10
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:584:5
     1: core::panicking::panic_fmt
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/panicking.rs:142:14
     2: core::result::unwrap_failed
               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/result.rs:1785:5
     3: core::result::Result<T,E>::unwrap
     4: build_script_build::main
     5: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</details>

File: xcommon/src/llvm.rs
Comment (dvc94ch) [reply xcommon/src/llvm.rs:49]: I think that's too fragile. Wondering about taking a different approach. How about we get the linker to tell us which libraries it needs from where? It looks like `ld.lld` supports a `--Map=filepath` argument that looks like it would give us just that.

**Comment by dvc94ch** [comment]
@MarijnS95 sorry for proposing all these last minute changes. Maybe if I'd have reviewed it earlier we could have been able to avoid it. To summarize my thoughs:

1. will merge #227 soon to prevent flutter regressions
2. detecting which libraries are needed could use the linker map instead of llvm-readobj + search path detection (search path detection seems error prone and incomplete)
3. all detected libraries that don't match a blacklist will be included
4. on android the blacklist is computed by searching the ndk's /usr/lib path 

I think this is the most generic approach and should easily generalize to the other platforms. Sorry for making you jump through hoops. Once it's merged I'm commited to maintaining it, so we need to get the basic design right.

**Comment by MarijnS95** [comment]
@dvc94ch Apologies for pretty much dropping the ball on this, been running into many other things that had higher prio and so this notification slipped to the end of my queue.

I totally understand where you're coming from and would love to do this differently too. The main pain-point being scanning various places for search paths and then perhaps not even maintaining the exact same order as `ld.lld`.

Anyway, since you approved this PR: Do you want to get this initial implementation in first, and let me rework it in a future PR - potentially removing lots of code again - or rather get that done first in this PR or a new PR? I'm already using this branch in automated Android CI builds to not be blocked, but I don't like being a downstream :)

**Comment by dvc94ch** [comment]
however for the time being if you fix the merge conflict I guess we can merge it.

**Comment by MarijnS95** [comment]
Let's see if I resolved that correctly :)

**Pull request merged**
